jwk/option.go

```diff
@@ -15,33 +15,58 @@ type identHTTPClient struct{}
 type identThumbprintHash struct{}
 type identRefreshInterval struct{}
 type identMinRefreshInterval struct{}
-type identRefreshBackoff struct{}
+type identFetchBackoff struct{}
 
-// WithHTTPClient allows users to specify the "net/http".Client object that
-// is used when fetching *jwk.Set objects.
-//
-// For historical reasons this method is also used in `jwk.Fetch*` functions,
-// eventhough the return value is marked as an `AutoRefreshOption`
-func WithHTTPClient(cl *http.Client) AutoRefreshOption {
-	return &autoRefreshOption{
-		option.New(identHTTPClient{}, cl),
-	}
+// AutoRefreshOption is a type of Option that can be passed to the
+// AutoRefresh object.
+type AutoRefreshOption interface {
+	Option
+	autoRefreshOption()
 }
 
-func WithThumbprintHash(h crypto.Hash) Option {
-	return option.New(identThumbprintHash{}, h)
+type autoRefreshOption struct {
+	Option
 }
 
-type autoRefreshOption struct {
+func (*autoRefreshOption) autoRefreshOption() {}
+
+// FetchOption is a type of Option that can be passed to `jwk.Fetch()`
+// This type also implements the `AutoRefreshOption`, and thus can be
+// safely passed to `(*jwk.AutoRefresh).Configure()`
+type FetchOption interface {
+	AutoRefreshOption
+	fetchOption()
+}
+
+type fetchOption struct {
 	Option
 }
 
-func (aro *autoRefreshOption) autoRefreshOption() bool {
-	return true
+func (*fetchOption) autoRefreshOption() {}
+func (*fetchOption) fetchOption()       {}
+
+// WithHTTPClient allows users to specify the "net/http".Client object that
+// is used when fetching jwk.Set objects.
+//
+func WithHTTPClient(cl *http.Client) FetchOption {
+	return &fetchOption{option.New(identHTTPClient{}, cl)}
+}
```

Comment:
It would be more extensible to accept the http client interface, which defines only Do instead of *http.Client.

```go
type Doer interface {
	Do(req *http.Request) (*http.Response, error)
}

func WithHTTPClient(cl Doer) FetchOption {
	return &fetchOption{option.New(identHTTPClient{}, cl)}
}
```

I sometimes use an http.Client compatible wrapper with Do (*http.Request, error) (*http.Response, error).

Reply:
> I sometimes use an http.Client compatible wrapper with Do (*http.Request, error) (*http.Response, error).

Huh, I didn't think about that possibility, but sure.  However, I don't think I will accept the name `Doer` because that doesn't explain anything to the person reading the document. It will have to be `type HTTPClient interface { Do(...) ... }`

Reply:
I know http.RoudTripper and that's great. My use case may be strange. You can think twice.

Reply:
Done